sample/marchingSquares.glsl

```diff
@@ -0,0 +1,204 @@
+#include "../generative/pnoise.glsl"
+#include "../math/const.glsl"
+#include "../math/rotate2d.glsl"
+#include "../sample.glsl"
+
+/*
+Author: Guido Schmidt
+description: The Marching Squares algorithm generates contour lines (isolines) of a given input for a
+given grid and a threshhold. Is uses a lookup of 16 cases to decide if a cell is inside, outside or
+a specific corner.
+use: <vec3> marchingSquares(in <vec2> uv) 
+options:
+    - SAMPLE_MARCHING_SQUARES_FNC(UV): optional sampling function, defaults so sampling of pnoise
+*/
+
+#ifndef SAMPLE_MARCHING_SQUARES_FNC
+float defaultSampleFunction(in vec2 uv) {
+    float dv = pnoise(uv * 5.0, vec2(0.0));
+    return dv;
+}
+#define SAMPLE_MARCHING_SQUARES_FNC(UV) defaultSampleFunction(UV)
+#endif
+
+#define centerOf(p1, p2, v1, v2) mix(p1, p2, 0.5)
+
+float sampleOutline(in vec2 p, in vec2 cellUv, in vec2 a, in vec2 b, in bool straight) {
+    float lineStrength = (straight ? 1.3333 : 1.0);
+    vec2 pa = p - a;
+    vec2 ba = b - a;
+    vec2 line = pa - ba * dot(pa, ba) / dot(ba, ba);
+    return 1.0 - step(lineStrength, length(line));
+}
+
+float sampleTile(in vec2 p, in vec2 cellUv, in vec2 a, in vec2 b, in int tile) {
+    vec2 pa = p - a;
+    vec2 ba = b - a;
+    vec2 line = pa - ba * dot(pa, ba) / dot(ba, ba);
+    switch (tile) {
+        case 0:
+          return step(0.5, cellUv.x);
+        case 1:
+          return step(0.5, 1.0 - cellUv.x);
+        case 2:
+          return step(0.5, 1.0 - cellUv.y);
+        case 3:
+          return step(0.5, cellUv.y);
+        case 4:
+          return 1.0 - step(0.35, (rotate2d(PI * 0.75) * cellUv).x);
+        case 5:
+          return step(0.35, (rotate2d(PI * 0.75) * cellUv).x);
+        case 6:
+          return 1.0 - step(1.05, (rotate2d(PI * 0.25) * cellUv).x);
+        case 7:
+          return step(1.05, (rotate2d(PI * 0.25) * cellUv).x);
+        case 8:
+          return step(0.35, (rotate2d(PI * 0.25) * cellUv).x);
+        case 9:
+          return 1.0 - step(0.35, (rotate2d(PI * 0.25) * cellUv).x);
+        case 10:
+          return 1.0 - step(0.35, (rotate2d(PI * -0.25) * cellUv).x);
+        case 11:
+          return step(0.35, (rotate2d(PI * -0.25) * cellUv).x);
+        case 12:
+          float shape12_0 = 1.0 - step(0.35, (rotate2d(PI * -0.25) * cellUv).x);
+          float shape12_1 = step(0.35, (rotate2d(PI * 0.75) * cellUv).x);
+          return shape12_0 - shape12_1;
+        case 13:
+          float shape13_0 = step(0.35, (rotate2d(PI * 0.25) * cellUv).x);
+          float shape13_1 = step(1.05, (rotate2d(PI * 0.25) * cellUv).x);
+          return shape13_0 - shape13_1;
+        default:
+          return 0.0;
+        // @DEBUG: enable the smoothstep in order to show a line for tiles
+        //return smoothstep(0.4, 0.0, length(line));
+    }
+}
+
+vec2 marchinSquares(in vec2 uv, in float cellSize, in float threshold, in vec2 resolution) {
```

Comment:
functions inside `sample` folder start with `sample` in this case would be `sampleMarchingSquares()`